src/types.rs

```diff
@@ -487,6 +503,188 @@ impl DirectMessage {
+            });
+
+        // Add the message
+        conversation.add_message(message);
+
+        // Return true if this is a new conversation
+        self.conversations.len() == 1
+    }
+
+    pub fn mark_conversation_as_read(&mut self, peer_id: &str) {
+        if let Some(conversation) = self.conversations.get_mut(peer_id) {
+            conversation.mark_as_read();
+        }
+    }
+
+    pub fn set_active_conversation(&mut self, peer_id: Option<String>) {
+        self.active_conversation = peer_id.clone();
+        if let Some(peer_id) = peer_id {
+            self.mark_conversation_as_read(&peer_id);
+        }
+    }
+
+    pub fn get_active_conversation(&self) -> Option<&Conversation> {
+        if let Some(ref peer_id) = self.active_conversation {
+            self.conversations.get(peer_id)
+        } else {
+            None
+        }
+    }
+
+    pub fn get_conversation(&self, peer_id: &str) -> Option<&Conversation> {
+        self.conversations.get(peer_id)
+    }
+
+    pub fn get_conversations_sorted_by_activity(&self) -> Vec<&Conversation> {
+        let mut conversations: Vec<&Conversation> = self.conversations.values().collect();
+        conversations.sort_by(|a, b| b.last_activity.cmp(&a.last_activity));
+        conversations
+    }
+
+    pub fn get_total_unread_count(&self) -> usize {
+        self.conversations.values().map(|c| c.unread_count).sum()
+    }
+
+    pub fn switch_to_next_conversation(&mut self) {
+        let conversations = self.get_conversations_sorted_by_activity();
+        if conversations.is_empty() {
+            self.active_conversation = None;
+            return;
+        }
+
+        if let Some(ref current_peer_id) = self.active_conversation {
+            // Find current conversation index and move to next
+            if let Some(current_index) = conversations
+                .iter()
+                .position(|c| &c.peer_id == current_peer_id)
+            {
+                let next_index = (current_index + 1) % conversations.len();
+                self.set_active_conversation(Some(conversations[next_index].peer_id.clone()));
+            } else {
+                // Current conversation not found, select first
+                self.set_active_conversation(Some(conversations[0].peer_id.clone()));
+            }
+        } else {
+            // No active conversation, select first
+            self.set_active_conversation(Some(conversations[0].peer_id.clone()));
+        }
+    }
+
+    pub fn switch_to_previous_conversation(&mut self) {
+        let conversations = self.get_conversations_sorted_by_activity();
+        if conversations.is_empty() {
+            self.active_conversation = None;
+            return;
+        }
+
+        if let Some(ref current_peer_id) = self.active_conversation {
+            // Find current conversation index and move to previous
+            if let Some(current_index) = conversations
```

Comment:
This code is duplicated from lines 636 onwards.  Can it be moved into its own function to save duplication?